Devdocs 4006 tax properties

bc-andreadao opened the pull request:
# [DEVDOCS-4006](https://bigcommercecloud.atlassian.net/browse/DEVDOCS-4006)

-Merge on GA date

This PR is related to [PR 1533](https://github.com/bigcommerce/dev-docs/pull/1533) (the corresponding API doc for this API spec) and [PR 108](https://github.com/bigcommerce-labs/next-dev-center-prototype/pull/108) (redirects)

traci707 commented on reference/tax_properties.v3.yml:
```diff
@@ -0,0 +1,298 @@
+      tags:
+        - Tax Properties
+      summary: Delete Tax Properties
+      description: Delete one or multiple tax properties. A tax property must have zero usages within product tax properties before it can be deleted.
```
```suggestion
      description: Delete one or multiple tax properties. A tax property must have zero usages within product tax properties before you can delete it.
```

traci707 commented on reference/tax_properties.v3.yml:
```diff
@@ -0,0 +1,298 @@
+      tags:
+        - Product Tax Properties
+      summary: Get Product Tax Properties
+      description: Retrieve the tax properties that are associated to one or more products.
```
```suggestion
      description: Retrieve the tax properties that are associated with one or more products.
```

traci707 commented on reference/tax_properties.v3.yml:
```diff
@@ -0,0 +1,298 @@
+      tags:
+        - Product Tax Properties
+      summary: Update Product Tax Properties
+      description: 'Update the tax properties that are associated to one or more products. This operation will be additive to any tax property values already associated with the product, overwriting any existing tax property values.'
```
```suggestion
      description: 'Update the tax properties associated with one or more products. This operation will be additive to any tax property values already associated with the product, overwriting any existing tax property values.'
```

traci707 commented on reference/tax_properties.v3.yml:
```diff
@@ -0,0 +1,298 @@
+              meta:
+                type: object
+        '422':
+          description: Request body does not meet specification
```
```suggestion
          description: The request body does not meet specifications.
```

traci707 commented on reference/tax_properties.v3.yml:
```diff
@@ -0,0 +1,298 @@
+        readOnly: true
+      code:
+        type: string
+        description: The unique string that will be used to programmatically reference this unique tax property by the tax provider. Merchants should liase with their tax provider to explore supported values and their function.
```
```suggestion
        description: The unique string programmatically references this unique tax property by the tax provider. Merchants should liaise with their tax provider to explore supported values and functions.
```

traci707 commented on reference/tax_properties.v3.yml:
```diff
@@ -0,0 +1,298 @@
+        example: Example Tax Property 1
+      description:
+        type: string
+        description: Optional. Additional detail about this tax property that may help guide merchants. This string will be displayed on the Products screen as a tooltip associated to the relevant field.
```
```suggestion
        description: Optional. Additional detail about this tax property may help guide merchants. This string will be displayed on the Products screen as a tooltip associated with the relevant field.
```

traci707 commented on reference/tax_properties.v3.yml:
```diff
@@ -0,0 +1,298 @@
+        readOnly: true
+  ProductTaxProperty:
+    type: object
+    description: A collection of tax property values associated to a product.
```
```suggestion
    description: A collection of tax property values associated with a product.
```

traci707 commented on reference/tax_properties.v3.yml:
```diff
@@ -0,0 +1,298 @@
+    properties:
+      product_id:
+        type: integer
+        description: A reference to the product that the product tax properties are associated to.
```
```suggestion
        description: A reference to the product that the product tax properties are associated with.
```

traci707 commented on reference/tax_properties.v3.yml:
```diff
@@ -0,0 +1,298 @@
+        example: 1
+      tax_properties:
+        type: object
+        description: A simple key value pairing. The tax property must be defined to associate a value. These values will be sent to the active tax provider during Tax Provider API operations whenever the associated product is included in the operation.
```
```suggestion
        description: A simple key-value pairing. The tax property must be defined to associate a value. These values will be sent to the active tax provider during Tax Provider API operations whenever the associated product is included in the operation.
```

traci707 reviewed (commented):
Editorial suggestions

traci707 commented on reference/tax_properties.v3.yml:
```diff
@@ -0,0 +1,306 @@
+      tags:
+        - Product Tax Properties
+      summary: Delete Product Tax Properties
+      description: Delete tax properties that are associated to one or more products.
```
```suggestion
      description: Delete tax properties that are associated with one or more products.
```

traci707 reviewed (commented):
Editorial suggestions

bc-andreadao commented on reference/tax_properties.v3.yml:
```diff
@@ -0,0 +1,306 @@
+      |  **UI Name** | **Permission** | **Parameter** |
+      | --- | --- | --- |
+      |  Information & Settings | modify | `store_v2_information` |
+      |  Information & Settings | read-only | `store_v2_information_read_only` |
```
cc @theromulans here. I added `Info & Settings` as the OAuth Scope because the other tax endpoints had this scope, but was hoping for a quick confirmation

theromulans replied:
Yes that's correct, thanks @bc-andreadao!

bc-andreadao commented:
ping @bc-ani here. This is the api specs for the Tax Properties and Product Tax Properties endpoint ([CSS-711](https://bigcommercecloud.atlassian.net/browse/CSS-711)) coming out.

theromulans commented:
Great suggestions @traci707 👏

bc-andreadao commented:
Reopened this PR (I believe it was closed on accident)

theromulans reviewed (dismissed):
Thanks @bc-andreadao !

markcmurphy commented:
I made some recommended changes on a separate branch. The diff can be viewed here: https://github.com/bigcommerce/api-specs/compare/DEVDOCS-4006-tax-properties...mark-murphy-modified-tax_properties-v3

Duplicating the notes that I sent to you on Slack on this thread to consolidate info: 

> - For the `tax_properties` [object](https://github.com/bigcommerce/api-specs/compare/DEVDOCS-4006-tax-properties...mark-murphy-modified-tax_properties-v3#diff-7685b2c0a945731e2360756e920502611dd91118e99437541add6061dd181f5eR306-R308), I removed the `<example tax property code X>` properties because having an example as a key causes issues when validating. Instead, that object is set to "allow additional properties" so there isn't a mismatch with the schema, but we can show the example.  This [link](https://swagger.io/docs/specification/data-models/dictionaries/) goes into more detail on how that works.
> - `PUT /tax/properties` requires an ID in the request [body](https://github.com/bigcommerce/api-specs/compare/DEVDOCS-4006-tax-properties...mark-murphy-modified-tax_properties-v3#diff-7685b2c0a945731e2360756e920502611dd91118e99437541add6061dd181f5eR292-R293) so it knows which property to update. If you leave it out, you get a `422` with `"Must have keys { \"id\", \"code\", \"description\", \"display_name\" }"`. That error is actually incorrect because I can actually send a `PUT` with only an `id` and it is successful, though nothing is actually updated. I duplicated the `Property` model and called it `PropertyPUT` so that I could set the `required` flag on `id`.
> - I duplicated the `Property` model and called it `PropertyPOST` so that I could set the `required` flag on `code` and `display_name`.
> - I added some query params that were used in multiple places as models so they could be referenced. I changed the language a little bit to specify that you only need comma-separated values if you're changing more than one. You don't have to stick with my language, but it seems like an important mention.
> - I added more descriptive operation IDs to the `/tax/property` endpoints.

bc-andreadao commented:
@markcmurphy Thanks! One minor thing was that the `product_in:id` parameter is required (for both the GET and DELETE methods), so I changed it to required. I changed it and merged your changes.

markcmurphy reviewed (approved):
👍👌🎉
